tests/jax/test_distributed_layernorm.py

```diff
@@ -0,0 +1,125 @@
+# Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# See LICENSE for license information.
+
+import pytest
+
+import jax
+import jax.numpy as jnp
+import numpy as np
+from jax import random
+from jax.sharding import Mesh, NamedSharding, PartitionSpec
+
+from distributed_test_base import generate_configs, generate_collectives_count
+from distributed_test_base import compare_ops
+from transformer_engine.jax import fp8_autocast
+from transformer_engine.jax.layernorm import layernorm
+
+DTYPES = [jnp.bfloat16, jnp.float32]
+
+
+class TestDistributedLayernorm:
+
+    def generate_inputs(self, shape, mesh_resource, dtype):
+        weight_shape = (shape[-1],)
+
+        x = random.normal(random.PRNGKey(1124), shape, dtype=dtype)
+        gamma = jnp.ones(weight_shape, dtype=dtype)
+        beta = jnp.ones(weight_shape, dtype=dtype)
+
+        if len(shape) == 2:
+            x_pspec = PartitionSpec(mesh_resource.dp_resource, None)
+        elif len(shape) == 3:
+            x_pspec = PartitionSpec(mesh_resource.dp_resource, None, None)
+        else:
+            raise NotImplementedError
+
+        g_pspec = b_pspec = PartitionSpec(None)
+
+        return (x, gamma, beta), (x_pspec, g_pspec, b_pspec)
+
+    def generate_collectives_count_ref(self, mesh_resource, ln_type):
+        is_dp_enabled = mesh_resource.dp_resource is not None
+        assert ln_type in ['layernorm', 'rmsnorm']
+        # for loss, dgamma and dbeta
+        ar_base_count = 3 if ln_type == 'layernorm' else 2
```

Comment:
3 all-reduce ops for `layernorm` type is 1 more than what I'm seeing with the previous xmap implementation (2). Is it my test script that's broken (see below) or is there a reason/bug for why the new layer norm is doing 1 extra collective?

```python
import numpy as np
from functools import partial

import jax
import jax.numpy as jnp
from jax import random
from jax.experimental.pjit import pjit, _UNSPECIFIED
from jax.sharding import PartitionSpec

from transformer_engine.jax.sharding import ShardingType, ShardingResource, global_shard_guard
from transformer_engine.jax.layernorm import layernorm

def get_sharding_spec(mesh_names, sharding_type):
    P = PartitionSpec
    if sharding_type is ShardingType.DP:
        return P(mesh_names[0], None), P(None), P(None)
    elif sharding_type is ShardingType.DP_TP_COL:
        return P(mesh_names[0], mesh_names[1]), P(None), P(None)
    else:
        raise NotImplementedError

def get_sharding_resource(mesh_names, sharding_type):
    dp_r = None
    tp_r = None
    
    if sharding_type in (ShardingType.DP, ShardingType.DP_TP_COL, ShardingType.DP_TP_ROW):
        dp_r = mesh_names[0]
    if sharding_type in (ShardingType.TP_COL, ShardingType.TP_ROW):
        tp_r = mesh_names[0]
    if sharding_type in (ShardingType.DP_TP_COL, ShardingType.DP_TP_ROW):
        tp_r = mesh_names[1]

    return ShardingResource(dp_r, tp_r)

def count_collectives(hlo):    
    tmp = hlo.splitlines()
    symb = "-start"
    result = {
        "all-reduce" : 0,
        "all-gather" : 0,
        "other" : 0
    }
    for line in tmp:
        txt = line.split()
        if len(txt) > 0 and symb in txt[0]:
            for collective in result.keys():
                if collective in txt[0]:
                    result[collective] += 1
    return result

def compare_ops(custom_func, ref_func,
                *args, grad_args=None,
                in_shardings=_UNSPECIFIED, out_shardings=_UNSPECIFIED,
                **kwargs):
    if isinstance(custom_func, partial):
        func_name = custom_func.func.__name__
    else:
        func_name = custom_func.__name__
    func_name = func_name.removeprefix('custom_')
    if grad_args is None:
        grad_args = tuple(range(len(args)))

    custom_gradded = jax.value_and_grad(custom_func, argnums=grad_args)
    test_fwd, test_grads = custom_gradded(*args, **kwargs)
    custom_pjitter = pjit(custom_gradded,
                            in_shardings=in_shardings,
                            out_shardings=out_shardings)
    custom_hlo = custom_pjitter.lower(*args, **kwargs).compile().as_text()
    custom_count = count_collectives(custom_hlo)

    print(f"`{func_name}`: Output collective count is {custom_count}.")

    ref_gradded = jax.value_and_grad(ref_func, argnums=grad_args)
    ref_fwd, ref_grads = ref_gradded(*args, **kwargs)
    fwd_tol = max(np.finfo(jnp.float16).eps**(2./3.), np.spacing(jnp.float16(ref_fwd)))
    assert jnp.allclose(test_fwd, ref_fwd, rtol=0.0, atol=fwd_tol), \
        f"`{func_name}`: Output (fwd) error {jnp.max(jnp.abs(test_fwd - ref_fwd))}" + \
        f" exceeds tolerance ({fwd_tol})."

    if len(grad_args) == 1:
        ref_grads = (ref_grads, )
        test_grads = (test_grads, )
    failed_grads = {}
    for i, grads in enumerate(zip(test_grads, ref_grads)):
        test_grad, ref_grad = grads
        if test_grad is None and ref_grad is None:
            continue
        bwd_tol = max(np.finfo(jnp.float32).eps**(2./3.),
                        np.spacing(jnp.max(jnp.abs(ref_grad)).astype(jnp.float32)))
        if not jnp.allclose(test_grad, ref_grad, rtol=0.0, atol=bwd_tol):
            failed_grads[i] = jnp.max(jnp.abs(test_grad - ref_grad))
    assert len(failed_grads) == 0, \
        f"`{func_name}`: Gradient (bwd) max errors" + \
        f" [{', '.join([f'Arg{k}={v}' for k,v in failed_grads.items()])}]" + \
        f" exceed tolerance ({bwd_tol})."


def custom_layernorm(x, gamma, beta, zero_centered_gamma, epsilon, sharding_type):
    result = layernorm(x, gamma, beta,
                        layernorm_type='layernorm',
                        zero_centered_gamma=zero_centered_gamma,
                        epsilon=epsilon,
                        sharding_type=sharding_type,
                        dp_dim_index=0)
    return jnp.mean(result)

def reference_layernorm(x, gamma, beta, zero_centered_gamma, epsilon, dtype=jnp.float16):
    x_ = jnp.asarray(x, jnp.float32)
    mean = jnp.mean(x_, axis=-1, keepdims=True)
    var = jnp.mean(jnp.square(x_ - mean), axis=-1, keepdims=True)
    normed_input = (x_ - mean) * jax.lax.rsqrt(var + epsilon)
    if zero_centered_gamma:
        result = jnp.asarray(normed_input * (gamma + 1) + beta).astype(dtype)
    else:
        result = jnp.asarray(normed_input * gamma + beta).astype(dtype)
    return jnp.mean(result)


def custom_rmsnorm(x, gamma, epsilon, sharding_type):
    result = layernorm(x, gamma, None,
                        layernorm_type='rmsnorm',
                        zero_centered_gamma=False,
                        epsilon=epsilon,
                        sharding_type=sharding_type,
                        dp_dim_index=0)
    return jnp.mean(result)

def reference_rmsnorm(x, gamma, epsilon, dtype=jnp.float16):
    x = jnp.asarray(x, jnp.float32)
    mean2 = jnp.mean(jax.lax.square(x), axis=-1, keepdims=True)
    y = jnp.asarray(x * jax.lax.rsqrt(mean2 + epsilon), dtype)
    result = y * gamma
    return jnp.mean(result)

num_gpus = 4
sharding_types = [ShardingType.DP, ShardingType.DP_TP_COL]
mesh_shapes = [(4, 1), (2, 2)]
mesh_names = [('dp', None), ('dp', 'tp')]
epsilon = 1e-5
zero_centered_gamma = False

dtype = jnp.float16
qkv_shape = (32, 128, 16, 64)
batch_size, _, num_heads, head_dim = qkv_shape
hidden_size = num_heads*head_dim
input_shape = (batch_size, hidden_size)
other_shape = (hidden_size, )

for sharding_type, mesh_shape, mesh_name in zip(sharding_types, mesh_shapes, mesh_names):
    print(f"Mesh Config: {sharding_type}, {mesh_shape}")
    custom_func = partial(custom_layernorm,
                          zero_centered_gamma=zero_centered_gamma,
                          epsilon=epsilon,
                          sharding_type=sharding_type)
    reference_func = partial(reference_layernorm,
                             zero_centered_gamma=zero_centered_gamma,
                             epsilon=epsilon, dtype=dtype)

    devices = np.asarray(jax.devices()[:num_gpus]).reshape(*mesh_shape)
    mesh = jax.sharding.Mesh(devices, mesh_name)
    with mesh, global_shard_guard(get_sharding_resource(mesh_name, sharding_type)):
        x_ = random.normal(random.PRNGKey(1124), input_shape, dtype=dtype)
        gamma_ = jnp.ones(other_shape, dtype=dtype)
        beta_ = jnp.ones(other_shape, dtype=dtype)
        compare_ops(custom_func, reference_func, x_, gamma_, beta_, grad_args=(0, 1, 2))

for sharding_type, mesh_shape, mesh_name in zip(sharding_types, mesh_shapes, mesh_names):
    print(f"Mesh Config: {sharding_type}, {mesh_shape}")
    custom_func = partial(custom_rmsnorm,
                          epsilon=epsilon,
                          sharding_type=sharding_type)
    reference_func = partial(reference_rmsnorm,
                             epsilon=epsilon, dtype=dtype)

    devices = np.asarray(jax.devices()[:num_gpus]).reshape(*mesh_shape)
    mesh = jax.sharding.Mesh(devices, mesh_name)
    with mesh, global_shard_guard(get_sharding_resource(mesh_name, sharding_type)):
        x_ = random.normal(random.PRNGKey(1124), input_shape, dtype=dtype)
        gamma_ = jnp.ones(other_shape, dtype=dtype)
        compare_ops(custom_func, reference_func, x_, gamma_, grad_args=(0, 1))
```

Reply:
The reason here is to set the upper bound of expected `allreduce`. For `layernorm`, it is expected to have at most three `allreduce`, 1 for loss, 1 for `dgamma` and 1 for `dbeta` and XLA fuses part of them, like fuse `dgamma` and `dbeta` as one. Therefore, the real number of collective ops is 2. However, the fusion strategy might change or be affected by `XLA_FLAGS` like `xla_gpu_all_reduce_combine_threshold_bytes=xxxx`, then `dgamma` and `dbeta` could be reduced by seperated `allreduce`. To make UTs be more robust to various  fusion strategy of XLA, I set the upper bound as reference collective op count, instead exact number.

I also have some ideas to make the checking be more general, like counting total bytes involved in collective ops, instead number of collective ops. We can submit a PR to improve it. Does it make sense?

Reply:
It is possible to detect those flags at runtime, but we might not need that. Since we use total bytes in collective ops as the criteria now, and the total bytes required communication is consistent across various fusion strategies of XLA.

For example, in layernorm, we have a loss in FP32[1] and `gamma`, and `beta` in BF16[1024,], and they require allreduce along DP. Then we examine if the total bytes of allreduce in HLO equals to 4 + 2*1024 + 2*1024 (FP32 + 1024*FP6 + 1024*FP6).

The implementation details could be found in [1f8d903](https://github.com/NVIDIA/TransformerEngine/pull/472/commits/1f8d9031d8657b259995e43c2e4773dfb672fe93#diff-597035bce1fbb5aa8a0af29199fdb714fa4a896f37af8c1c8fd58212dfb4f09a).

Would you kindly review it to see if it is acceptable? Thanks